### PR TITLE
Feature/review page refactor#10114

### PR DESCRIPTION
**Pull request by Clark-W-Griswold**
## Description
Refactoring review page:
 - Start to change component composition.
 - Start to move towards less prop drilling. 

## Testing done
Unit tests updated to pass with new code. 
ensured e2e tests were passing as expected. 

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/platform/forms/components/common/grid/Column.jsx
Comment (cvalarida) [inline src/platform/forms/components/common/grid/Column.jsx:10]: Oh, cool. Didn't realize this was a thing. :+1: 

My understanding is that it'll only re-render when the props change, but I'm not sure how this is different from a standard functional component. What benefit do we gain here? (Asking to sate my own curiosity more than anything else.)

File: src/platform/forms/components/review/ApplicationSubmitted.jsx
Comment (cvalarida) [inline src/platform/forms/components/review/ApplicationSubmitted.jsx:20]: Won't this render `<div classname="columns columns"/>`?

File: src/platform/forms/containers/review/PreSubmitSection.jsx
Comment (cvalarida) [inline src/platform/forms/containers/review/PreSubmitSection.jsx:20]: Where does `showPresubmitError` come from?

File: src/platform/forms/components/review/ClientError.jsx
Comment (cvalarida) [inline src/platform/forms/components/review/ClientError.jsx:5]: What do you think about putting an `index.js` in `common/grid/` so we can
```js
import { Column, Row } from 'platform/forms/components/common/grid';
```
?

File: src/platform/forms/components/review/SubmitButtons.jsx
Comment (cvalarida) [inline src/platform/forms/components/review/SubmitButtons.jsx:19]: Why is this empty? In the [old version](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/review/SubmitButtons.jsx#L158-L160), this rendered the submit message.
```js
      <div className="row">
        <div className="columns">{submitMessage}</div>
      </div>
```

File: src/platform/forms/components/review/SubmitButtons.jsx
Comment (cvalarida) [inline src/platform/forms/components/review/SubmitButtons.jsx:39]: I see this is in the old version as well, so I won't suggest changing anything for now, but in a future iteration, we should figure out why this is here if it's actually hidden.

File: src/platform/forms/components/review/SubmitPending.jsx
Comment (cvalarida) [inline src/platform/forms/components/review/SubmitPending.jsx:21]: Why?

File: src/platform/forms/components/review/GenericError.jsx
Comment (cvalarida) [inline src/platform/forms/components/review/GenericError.jsx:63]: Is this fragment needed?

**Review by cvalarida — COMMENTED** [review]
All the comments! And more that didn't fit nicely into inline review comments:
1. I'm uncomfortable with the lack of automated testing combined with pushing this out into production immediately.
1. I made a lot of comments about prop types, but eventually stopped. :stuck_out_tongue: I think it'll be helpful for future refactoring to be really specific with the prop types, but I trust your discretion there.
1. `SubmitButtons`, `SubmitPending`, `ThrottledError`, `ValidationError` all have the same basic layout. Is it worth abstracting that?

File: src/platform/forms/save-in-progress/helpers.js
Comment (Clark-W-Griswold) [reply src/platform/forms/save-in-progress/helpers.js:8]: That is correct. I am open to using flipper to control the toggle or exploring other options if we would like.

File: src/platform/forms/containers/review/RoutedSavableReviewPage.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/containers/review/RoutedSavableReviewPage.jsx:49]: Good call. I'm going to extract things like this so we can create tickets around the effort.

File: src/platform/forms/containers/review/RoutedSavableReviewPage.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/containers/review/RoutedSavableReviewPage.jsx:59]: I think we should absolutely re-evaluate this in a future iteration. For now, I was just trying to mirror our logic here:
https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms/save-in-progress/RoutedSavableReviewPage.jsx#L49

File: src/platform/forms/components/review/ClientError.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/ClientError.jsx:5]: Love this idea and I think we will move more towards that pattern as we become a library. 😃

File: src/platform/forms/components/review/SubmitButtons.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/SubmitButtons.jsx:19]: Correct, but for the case in which we are just rendering the buttons `(submission.status === false)` and not setting the `submitMessage`. So in respect to keeping the same html footprint as prior to the refactor, we have that there. 

`let submitMessage;`

File: src/platform/forms/components/review/SubmitButtons.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/SubmitButtons.jsx:39]: I completely agree and I think this is somewhat related to the empty column above. I left it that way in this iteration as it felt out the scope of the refactor.

File: src/platform/forms/components/review/SubmitPending.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/SubmitPending.jsx:21]: Just keeping with the spirit of the previous code. Same as above where `submitMessage` was never set and therefore undefined.

File: src/platform/forms/components/review/GenericError.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/GenericError.jsx:63]: It is not. Good catch. 😃

**Comment by Clark-W-Griswold** [comment]
> All the comments! And more that didn't fit nicely into inline review comments:
> 
> 1. I'm uncomfortable with the lack of automated testing combined with pushing this out into production immediately.
> 2. I made a lot of comments about prop types, but eventually stopped. 😛 I think it'll be helpful for future refactoring to be really specific with the prop types, but I trust your discretion there.
> 3. `SubmitButtons`, `SubmitPending`, `ThrottledError`, `ValidationError` all have the same basic layout. Is it worth abstracting that?

1. Going to push a follow up commit to cast a wider net on coverage. Will send a follow up message when that is ready.
2. Agreed. I 100% want to layer in that in the short term. If I can get to that tonight with the above testing, expect that in this PR. 😄 
3. 100% agreed. I think we can create a generic component here. I can create a ticket to follow up with this.

File: src/platform/forms/containers/review/SubmitController.jsx
Comment (rianfowler) [inline src/platform/forms/containers/review/SubmitController.jsx:2]: summarize changes made to this file

File: src/platform/forms/containers/review/SubmitController.jsx
Comment (rianfowler) [reply src/platform/forms/containers/review/SubmitController.jsx:111]: this should be in the enum if we're rewriting this switch statement

File: src/platform/forms/containers/review/SubmitController.jsx
Comment (rianfowler) [reply src/platform/forms/containers/review/SubmitController.jsx:122]: add `TODO: routing`

File: src/platform/forms/save-in-progress/helpers.js
Comment (rianfowler) [reply src/platform/forms/save-in-progress/helpers.js:8]: What's your rollback plan?

**Review by rianfowler — COMMENTED** [review]
This change needs more explanation. It's good as the contributor of such a large change to do your own review to draw attention to the important changes or have instructions on how people should review it. It's hard to tell what's new vs old logic in this since nothing has been deleted. 

I added some name change suggestions but I think we could talk it over with the team. I don't like terse names but we should come to a consensus on how we want to name things. 

These changes need a test plan. Since we're making layout changes, I suggest getting a review from Trevor.

File: src/platform/forms/containers/review/SubmitController.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/containers/review/SubmitController.jsx:111]: Sounds good to me. This is a part of this code that I'd like to come back to at some point and find a better approach.

File: src/platform/forms/containers/review/PreSubmitSection.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/containers/review/PreSubmitSection.jsx:18]: This was a copy pasta from this:
https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/review/SubmitController.jsx#L80

Maybe we can update the description for now and circle back on this component in a future iteration?

File: src/platform/forms/selectors/review/index.js
Comment (Clark-W-Griswold) [reply src/platform/forms/selectors/review/index.js:6]: For composition mostly

Comment (Clark-W-Griswold) [reply src/platform/forms/selectors/review/index.js:6]: showLoginModalSelector is an example of this.

File: src/platform/forms/containers/review/SubmitController.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/containers/review/SubmitController.jsx:179]: useEffect will take a second param with options of the props/things to evaluate. If changed, it will trigger. Component will always return the activeComponent as set by useState (default) or by the  setState method returned by useState (defined as setActiveComponent here).

File: src/platform/forms/components/common/grid/Column.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/common/grid/Column.jsx:10]: Probably not the best place for them and going to remove for now. Maybe we revisit and layer in as needed on some of the larger components?

File: src/platform/forms/components/review/ApplicationSubmitted.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/ApplicationSubmitted.jsx:20]: not needed here. that is me not properly cleaning up. 😃

File: src/platform/forms/containers/review/PreSubmitSection.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/containers/review/PreSubmitSection.jsx:20]: `form?.submission?.hasAttemptedSubmit`

will update the proptypes 😃

File: src/platform/forms/components/review/SubmitPending.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/SubmitPending.jsx:21]: cleaned up className

File: src/platform/forms/components/review/ApplicationSubmitted.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/ApplicationSubmitted.jsx:32]: This is keeping the existing classes that were on the div prior. The new component will append classNames passed to it. 

Should we migrate now or create a ticket for a future iteration?

File: src/platform/forms/components/review/ApplicationSubmitted.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/ApplicationSubmitted.jsx:18]: Just keeping with what was there before from a markup perspective. 
https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/review/SubmitButtons.jsx#L140
https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/review/SubmitButtons.jsx#L157

Difference here is the child components of the previously existing `React.Fragment` component. Mainly the `Column` and `Row` reusable components and the overall composed template for an `applicationSubmitted` message

File: src/platform/forms/components/review/ApplicationSubmitted.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/ApplicationSubmitted.jsx:1]: These comments are simply my attempt to organize the imports in an effort to clean up and make them more refactor-friendly. Let me know your thoughts. Happy to expand further if needed. 😃

File: src/platform/forms/components/review/GenericError.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/GenericError.jsx:2]: This component comes from the `renderErrorMessage` function that was passed down the component tree. To follow more of a compositional pattern, I broke this component out into a reusable version of what it was before.

Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/GenericError.jsx:2]: Here is the previous version in the `RoutedSavableReviewPage`:
https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms/save-in-progress/RoutedSavableReviewPage.jsx#L63

File: src/platform/forms/components/review/ThrottledError.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/components/review/ThrottledError.jsx:19]: sounds good to me 👍

File: src/platform/forms/containers/review/SubmitController.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/containers/review/SubmitController.jsx:44]: Updated constants and components as needed. pulled changes accepted in this PR and iterating through all tests and linting.

File: src/platform/forms/containers/review/RoutedSavableReviewPage.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/containers/review/RoutedSavableReviewPage.jsx:1]: This is a first attempt at breaking down the logic in this file. We've sorted the imports and started to remove functions in the classes and break them out into reusable components.

File: src/platform/forms/containers/review/SubmitController.jsx
Comment (Clark-W-Griswold) [reply src/platform/forms/containers/review/SubmitController.jsx:2]: This is an attempt to reduce the previous `SubmitController` and `SubmitButtons` components. We are moving all of the logic of what is shown into the `SubmitController` and allowing the components it renders to be their own, re-usable versions.

File: src/platform/forms/components/review/ApplicationSubmitted.jsx
Comment (rianfowler) [reply src/platform/forms/components/review/ApplicationSubmitted.jsx:32]: I'm not sure- I think it's desirable but I don't know how feasible it is or what we gain from it.

**Review by rianfowler — APPROVED** [review]

**Pull request closed without merging**
